File: May 2022/May_19.cpp
class Solution {
public:
    int longestIncreasingPath(vector<vector<int>>& matrix) {
        int n = matrix.size(), m = matrix[0].size();
        queue<pair<int, int> > q;
        for(int i = 0; i < n; i++)
            for(int j = 0; j < m; j++)
                q.push({i, j});
        
        int dirs[4][2] = {{-1, 0}, {0, -1}, {1, 0}, {0, 1}};
        int dp[n][m];
        memset(dp, 0, sizeof(dp));
        int t = 1;
        while(!q.empty()){
            int size = q.size();
            while(size--){
                int i = q.front().first, j = q.front().second;
                q.pop();
                if(dp[i][j] >= t)
                    continue;
                dp[i][j] = t;
                for(auto &d : dirs){
                    int x = i + d[0], y = j + d[1];
                    if(x >= 0 && y >= 0 && x < n && y < m && matrix[x][y] > matrix[i][j])
                        q.push({x, y});
                }
            }
            t++;
        }
        return t - 1;
    }
};

File: May 2022/May_19.java
class Solution {
    public int longestIncreasingPath(int[][] matrix) {
        int n = matrix.length, m = matrix[0].length;
        
        Queue<int[]> q = new ArrayDeque<>();
        for(int i = 0; i < n; i++)
            for(int j = 0; j < m; j++)
                q.add(new int[]{i, j});
        
        int[][] dirs = {{-1, 0}, {0, -1}, {1, 0}, {0, 1}};
        int[][] dp = new int[n][m];
        int t = 1;
        while(!q.isEmpty()){
            int size = q.size();
            while(size-- > 0){
                int[] rem = q.remove();
                if(dp[rem[0]][rem[1]] >= t)
                    continue;
                dp[rem[0]][rem[1]] = t;
                
                for(int[] d : dirs){
                    int x = rem[0] + d[0], y = rem[1] + d[1];
                    if(x >= 0 && y >= 0 && x < n && y < m && matrix[x][y] > matrix[rem[0]][rem[1]])
                        q.add(new int[]{x, y});
                }
            }
            t++;
        }
        return t - 1;
    }
}
